### ISSUE TITLE
Job module run on Master tenant only

### ISSUE THREAD

**Issue by lthoa** [oqtane/oqtane.framework#1628]
When we install sites on separate databases then the Job table just created on Master tenant database and Job module execute on the master site only.

**Comment by lthoa** [comment]
But when calling siteRepository.GetSites() it only returns the sites on the Master database, then only the Notifications on the Master database are processed.

**Comment by sbwalker** [comment]
I will have to investigate. Scheduled jobs do not run in a web context so the SiteState needs to be explicitly set for tenancy purposes. The code which executes the job is setting the SiteState so I am not sure why it is not working for you.

**Comment by sbwalker** [comment]
@lthoa are you running on Blazor WebAssembly or Blazor Server?

**Comment by lthoa** [comment]
I'm running on Server mode.

The JobLog shows Succeeded for Default Site only. Other sites are not executed

![image](https://user-images.githubusercontent.com/6874322/131360865-c28e17f6-3cfc-46b0-865b-a2f1e501fa53.png)

**Comment by sbwalker** [comment]
I am able to reproduce the problem when using tenants with separate databases. Its very confusing as the SiteState.TenantId is being set properly by HostedServiceBase, and I can verify SiteState is correct within ExecuteJob... however SiteState seems to be lost when it connects to the database. The interesting thing I noticed is that the SiteRepostory.GetSites() method gets called twice - once for each Tenant - which is correct, however the DBContextBase.OnConfiguring() method is only called once... so this would seem to indicate that there is a shared service scope - which should not be the case. I will need to continue investigating.

**Comment by sbwalker** [comment]
So it appears that DBContext is scoped to a request... and IHostedService is not scoped to a request. So when you have multiple tenants - each with their own database - when a scheduled job executes and calls SiteRepostory.GetSites() a single DBContext is created and used for the duration of that scheduled job. This is a problem as it means that DBContextBase.OnConfiguring() method is not called for each tenant - only the first tenant. So this explains the behavior you are seeing. I believe the solution is to use a DBContext factory - however this will require further investigation.

**Comment by sbwalker** [comment]
There is a way to override the default behavior of AddDbContext so that it is used Transient scope rather than Scoped(Request) scope:

```
        public static IServiceCollection AddOqtaneDbContext(this IServiceCollection services)
        {
            services.AddDbContext<MasterDBContext>(options => { }, ServiceLifetime.Transient);
            services.AddDbContext<TenantDBContext>(options => { }, ServiceLifetime.Transient);

            return services;
        }
```

After making this change, the scheduled job works as expected for all tenants (using multiple databases):

![image](https://user-images.githubusercontent.com/4840590/131499825-8aefde55-fafc-49c5-a377-18576ca77838.png)

So this revolves the problem... but more more investigation is needed to determine if there are any other side effects of this change. I do not believe there should be as the API and Repository services are transient as well.

**Comment by sbwalker** [comment]
After significant testing of various scenarios I believe using a Transient DBContext is the appropriate solution for this issue - it is fixed in #1637. This fix will actually help a few other scenarios as well. Thank you very much for reporting this issue!
